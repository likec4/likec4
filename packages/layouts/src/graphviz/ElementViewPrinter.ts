import type { ComputedEdge, ComputedElementView, Fqn } from '@likec4/core'
import { compareFqnHierarchically, DefaultArrowType, defaultTheme as Theme, nonNullable } from '@likec4/core'
import { filter, first, isTruthy, last, map, pipe, reverse, sort } from 'remeda'
import type { EdgeModel, RootGraphModel } from 'ts-graphviz'
import { attribute as _ } from 'ts-graphviz'
import { edgeLabel } from './dot-labels'
import { DefaultEdgeStyle, DotPrinter } from './DotPrinter'
import type { DotSource } from './types'
import { isCompound, toArrowType } from './utils'

export class ElementViewPrinter extends DotPrinter<ComputedElementView> {
  static toDot(view: ComputedElementView): DotSource {
    return new ElementViewPrinter(view).print()
  }

  protected override buildGraphvizModel(G: RootGraphModel): void {
    super.buildGraphvizModel(G)

    const groups = pipe(
      this.view.nodes,
      filter(n => n.children.length > 0),
      map(n => n.id),
      sort(compareFqnHierarchically),
      reverse(),
      map(id => {
        // edges only inside clusters, compound endpoints are not considered
        const edges = this.findNestedEdges(id).filter(e =>
          !!this.getGraphNode(e.source) && !!this.getGraphNode(e.target)
        )
        return { id, edges }
      }),
      filter(({ edges }) => edges.length > 1 && edges.length <= 8)
    )

    const processed = new Set<Fqn>()
    for (const group of groups) {
      const edges = group.edges.filter(e => !processed.has(e.source) && !processed.has(e.target))
      for (const edge of edges) {
        try {
          const sourceNode = nonNullable(this.getGraphNode(edge.source), `Graphviz Node not found for ${edge.source}`)
          const targetNode = nonNullable(this.getGraphNode(edge.target), `Graphviz Node not found for ${edge.target}`)
          processed.add(edge.source)
          processed.add(edge.target)
          sourceNode.attributes.set(_.group, group.id)
          targetNode.attributes.set(_.group, group.id)
        } catch (e) {
          console.error(e)
        }
      }
    }
  }

  protected override addEdge(edge: ComputedEdge, G: RootGraphModel): EdgeModel | null {
    const viewEdges = this.view.edges
    const [sourceFqn, targetFqn] = edge.dir === 'back' ? [edge.target, edge.source] : [edge.source, edge.target]
    const [sourceNode, source, ltail] = this.edgeEndpoint(sourceFqn, nodes => last(nodes))
    const [targetNode, target, lhead] = this.edgeEndpoint(targetFqn, first)

    const edgeParentId = edge.parent
    const parent = edgeParentId === null
      ? G
      : nonNullable(this.getSubgraph(edgeParentId), `Parent not found for edge ${edge.id}`)

    const e = parent.edge([source, target], {
      [_.likec4_id]: edge.id,
      [_.style]: edge.line ?? DefaultEdgeStyle
    })

    lhead && e.attributes.set(_.lhead, lhead)
    ltail && e.attributes.set(_.ltail, ltail)

    if (lhead || ltail) {
      const sourceId = source.attributes.get(_.likec4_id) as Fqn
      const targetId = target.attributes.get(_.likec4_id) as Fqn
      const existingVisibleEdge = viewEdges.find(e => e.source === sourceId && e.target === targetId)
      if (existingVisibleEdge) {
        e.attributes.set(_.weight, 0)
      }
    }

    const label = isTruthy(edge.label) ? edgeLabel(edge.label) : null
    if (isTruthy(label)) {
      if (lhead || ltail) {
        e.attributes.set(_.xlabel, label)
      } else {
        e.attributes.set(_.label, label)
      }
    }
    if (edge.color) {
      e.attributes.apply({
        [_.color]: Theme.relationships[edge.color].lineColor,
        [_.fontcolor]: Theme.relationships[edge.color].labelColor
      })
    }

    let [head, tail] = [edge.head ?? DefaultArrowType, edge.tail ?? 'none']

    if (edge.dir === 'back') {
      e.attributes.apply({
        [_.arrowtail]: toArrowType(head),
        [_.dir]: 'back'
      })
      if (tail !== 'none') {
        e.attributes.apply({
          [_.arrowhead]: toArrowType(tail),
          [_.dir]: 'both'
        })
      }
      return e
    }

    if (head === 'none' && tail === 'none') {
      e.attributes.apply({
        [_.arrowtail]: 'none',
        [_.arrowhead]: 'none',
        [_.dir]: 'none',
        [_.constraint]: false
      })
      return e
    }

    if (head !== 'none' && tail !== 'none') {
      e.attributes.apply({
        [_.arrowhead]: toArrowType(head),
        [_.arrowtail]: toArrowType(tail),
        [_.dir]: 'both',
        [_.minlen]: 0
      })
      return e
    }

    if (head !== DefaultArrowType) {
      e.attributes.set(_.arrowhead, toArrowType(head))
    }
    if (tail !== 'none') {
      e.attributes.set(_.arrowtail, toArrowType(tail))
    }

    // This heuristic removes the rank constraint from the edge
    // if it is the only edge within container.
    let otherEdges
    if (edgeParentId === null && sourceNode.parent == null && targetNode.parent == null) {
      otherEdges = viewEdges.filter(e => {
        // hide self
        if (e.id === edge.id) {
          return false
        }
        // hide edges with the same endpoints
        if (
          (e.source === edge.source && e.target === edge.target)
          || (e.source === edge.target && e.target === edge.source)
        ) {
          return false
        }
        // hide edges inside clusters
        if (e.parent !== null) {
          return false
        }
        const edgeSource = this.viewElement(e.source)
        const edgeTarget = this.viewElement(e.target)
        // hide edges with compound endpoints
        if (isCompound(edgeSource) || isCompound(edgeTarget)) {
          return false
        }
        // only edges between top-level nodes
        return edgeSource.parent == null && edgeTarget.parent == null
      })
    } else {
      otherEdges = this.findNestedEdges(edgeParentId).filter(e => {
        // hide self
        if (e.id === edge.id) {
          return false
        }
        // hide edges with the same endpoints
        if (
          (e.source === edge.source && e.target === edge.target)
          || (e.source === edge.target && e.target === edge.source)
        ) {
          return false
        }
        return true
      })
    }
    const isTheOnlyEdge = otherEdges.length === 0
    if (isTheOnlyEdge) {
      if (edgeParentId === null || this.leafElements(edgeParentId).length <= 3) {
        // don't rank the edge
        // e.attributes.set(_.minlen, 0)
        e.attributes.set(_.constraint, false)
      }
    }

    return e
  }
}
