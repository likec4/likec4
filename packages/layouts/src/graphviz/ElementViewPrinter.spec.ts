import type { ComputedElementView } from '@likec4/core'
import { describe, it } from 'vitest'
import {
  computedAmazonView,
  computedCloud3levels,
  computedCloudView,
  computedIndexView,
  parsedModel,
} from './__fixtures__'
import { ElementViewPrinter } from './ElementViewPrinter'

function print(view: ComputedElementView) {
  return new ElementViewPrinter(view, parsedModel.specification).print()
}

describe('ElementViewPrinter', () => {
  it('computedIndexView', async ({ expect }) => {
    const diagram = print(computedIndexView)
    await expect(diagram).toMatchFileSnapshot('__snapshots__/ElementViewPrinter-index.dot')
  })
  it('computedAmazonView', async ({ expect }) => {
    const diagram = print(computedAmazonView)
    await expect(diagram).toMatchFileSnapshot('__snapshots__/ElementViewPrinter-amazon.dot')
  })

  it('computedCloud3levels', async ({ expect }) => {
    const diagram = print(computedCloud3levels)
    await expect(diagram).toMatchFileSnapshot('__snapshots__/ElementViewPrinter-cloud3levels.dot')
  })

  it('computedCloudView', async ({ expect }) => {
    const diagram = print(computedCloudView)
    await expect(diagram).toMatchFileSnapshot('__snapshots__/ElementViewPrinter-cloud.dot')
  })
})
