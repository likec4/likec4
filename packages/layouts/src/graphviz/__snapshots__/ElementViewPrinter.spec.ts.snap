// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ElementViewPrinter > computedAmazonView 1`] = `
"digraph {
  bgcolor = "transparent";
  layout = "dot";
  compound = true;
  rankdir = "TB";
  TBbalance = "min";
  splines = "spline";
  outputorder = "nodesfirst";
  nodesep = 1.528;
  ranksep = 1.667;
  pack = 120;
  packmode = "array_3";
  pad = 0.209;
  fontname = "Arial";
  graph [
    fontsize = 15;
    labeljust = "l";
    labelloc = "t";
    margin = 50.1;
  ];
  edge [
    arrowsize = 0.75;
    fontname = "Arial";
    fontsize = 14;
    penwidth = 2;
    color = "#6E6E6E";
    fontcolor = "#C6C6C6";
  ];
  node [
    fontname = "Arial";
    shape = "rect";
    width = 4.445;
    height = 2.5;
    style = "filled";
    penwidth = 0;
  ];
  subgraph "cluster_cloud" {
    likec4_id = "cloud";
    likec4_level = 0;
    likec4_depth = 1;
    fillcolor = "#194b9e";
    color = "#1b3d88";
    style = "filled";
    margin = 32;
    label = <<FONT POINT-SIZE="11" COLOR="#bfdbfeb3"><B>CLOUD</B></FONT>>;
    "backend" [
      likec4_id = "cloud.backend";
      likec4_level = 1;
      fillcolor = "#3b82f6";
      fontcolor = "#eff6ff";
      color = "#2563eb";
      margin = "0.362,0.362";
      label = <<FONT POINT-SIZE="19">Backend</FONT>>;
    ];
  }
  subgraph "cluster_amazon" {
    likec4_id = "amazon";
    likec4_level = 0;
    likec4_depth = 1;
    fillcolor = "#194b9e";
    color = "#1b3d88";
    style = "filled";
    margin = 32;
    label = <<FONT POINT-SIZE="11" COLOR="#bfdbfeb3"><B>AMAZON</B></FONT>>;
    "s3" [
      likec4_id = "amazon.s3";
      likec4_level = 1;
      fillcolor = "#3b82f6";
      fontcolor = "#eff6ff";
      color = "#2563eb";
      margin = "0.362,0.362";
      label = <<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="6"><TR><TD><FONT POINT-SIZE="19">S3</FONT></TD></TR><TR><TD><FONT POINT-SIZE="14" COLOR="#bfdbfe">S3 is a storage service</FONT></TD></TR></TABLE>>;
    ];
  }
  "backend" -> "s3" [
    likec4_id = "cloud.backend:amazon.s3";
    style = "dashed";
    weight = 1;
    label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">persists artifacts</FONT></TD></TR></TABLE>>;
    arrowhead = "normal";
    arrowtail = "odiamond";
    dir = "both";
    minlen = 0;
  ];
}"
`;

exports[`ElementViewPrinter > computedCloud3levels 1`] = `
"digraph {
  bgcolor = "transparent";
  layout = "dot";
  compound = true;
  rankdir = "TB";
  TBbalance = "min";
  splines = "spline";
  outputorder = "nodesfirst";
  nodesep = 1.528;
  ranksep = 1.667;
  pack = 120;
  packmode = "array_3";
  pad = 0.209;
  fontname = "Arial";
  graph [
    fontsize = 15;
    labeljust = "l";
    labelloc = "t";
    margin = 50.1;
  ];
  edge [
    arrowsize = 0.75;
    fontname = "Arial";
    fontsize = 14;
    penwidth = 2;
    color = "#6E6E6E";
    fontcolor = "#C6C6C6";
  ];
  node [
    fontname = "Arial";
    shape = "rect";
    width = 4.445;
    height = 2.5;
    style = "filled";
    penwidth = 0;
  ];
  "support" [
    likec4_id = "support";
    likec4_level = 0;
    fillcolor = "#3b82f6";
    fontcolor = "#eff6ff";
    color = "#2563eb";
    margin = "0.362,0.362";
    label = <<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="6"><TR><TD><FONT POINT-SIZE="19">Support Engineer</FONT></TD></TR><TR><TD><FONT POINT-SIZE="14" COLOR="#bfdbfe">Support engineers are responsible for<BR/>supporting customers</FONT></TD></TR></TABLE>>;
  ];
  "customer" [
    likec4_id = "customer";
    likec4_level = 0;
    fillcolor = "#3b82f6";
    fontcolor = "#eff6ff";
    color = "#2563eb";
    margin = "0.362,0.362";
    label = <<FONT POINT-SIZE="19">customer</FONT>>;
  ];
  "amazon" [
    likec4_id = "amazon";
    likec4_level = 0;
    fillcolor = "#3b82f6";
    fontcolor = "#eff6ff";
    color = "#2563eb";
    margin = "0.362,0.362";
    label = <<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="6"><TR><TD><FONT POINT-SIZE="19">Amazon</FONT></TD></TR><TR><TD><FONT POINT-SIZE="14" COLOR="#bfdbfe">Amazon is a cloud provider</FONT></TD></TR></TABLE>>;
  ];
  subgraph "cluster_cloud" {
    likec4_id = "cloud";
    likec4_level = 0;
    likec4_depth = 2;
    fillcolor = "#1a468d";
    color = "#1c3979";
    style = "filled";
    margin = 40;
    label = <<FONT POINT-SIZE="11" COLOR="#bfdbfeb3"><B>CLOUD</B></FONT>>;
    "adminpanel" [
      likec4_id = "cloud.frontend.adminPanel";
      likec4_level = 1;
      fillcolor = "#3b82f6";
      fontcolor = "#eff6ff";
      color = "#2563eb";
      margin = "0.362,0.362";
      label = <<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="6"><TR><TD><FONT POINT-SIZE="19">Admin Panel Webapp</FONT></TD></TR><TR><TD><FONT POINT-SIZE="14" COLOR="#bfdbfe">The admin panel is a webapp that allows<BR/>support staff to manage customer data.</FONT></TD></TR></TABLE>>;
      group = "cloud";
    ];
    "dashboard" [
      likec4_id = "cloud.frontend.dashboard";
      likec4_level = 1;
      fillcolor = "#3b82f6";
      fontcolor = "#eff6ff";
      color = "#2563eb";
      margin = "0.362,0.362";
      label = <<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="6"><TR><TD><FONT POINT-SIZE="19">Customer Dashboard Webapp</FONT></TD></TR><TR><TD><FONT POINT-SIZE="14" COLOR="#bfdbfe">The customer dashboard is a webapp that<BR/>allows customers to view their data.</FONT></TD></TR></TABLE>>;
      group = "cloud";
    ];
    subgraph "cluster_backend" {
      likec4_id = "cloud.backend";
      likec4_level = 1;
      likec4_depth = 1;
      fillcolor = "#194b9e";
      color = "#1b3d88";
      style = "filled";
      margin = 40;
      label = <<FONT POINT-SIZE="11" COLOR="#bfdbfeb3"><B>BACKEND</B></FONT>>;
      "graphql" [
        likec4_id = "cloud.backend.graphql";
        likec4_level = 2;
        fillcolor = "#3b82f6";
        fontcolor = "#eff6ff";
        color = "#2563eb";
        margin = "0.362,0.362";
        label = <<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="6"><TR><TD><FONT POINT-SIZE="19">Graphql API</FONT></TD></TR><TR><TD><FONT POINT-SIZE="14" COLOR="#bfdbfe">Component that allows to query data via<BR/>GraphQL.</FONT></TD></TR></TABLE>>;
        group = "cloud";
      ];
      "storage" [
        likec4_id = "cloud.backend.storage";
        likec4_level = 2;
        fillcolor = "#3b82f6";
        fontcolor = "#eff6ff";
        color = "#2563eb";
        margin = "0.362,0";
        penwidth = 2;
        shape = "cylinder";
        label = <<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="6"><TR><TD><FONT POINT-SIZE="19">Backend Storage</FONT></TD></TR><TR><TD><FONT POINT-SIZE="14" COLOR="#bfdbfe">The backend storage is a component that<BR/>stores data.</FONT></TD></TR></TABLE>>;
        group = "cloud";
      ];
      "graphql" -> "storage" [
        likec4_id = "cloud.backend.graphql:cloud.backend.storage";
        style = "dashed";
        weight = 4;
        minlen = 0;
      ];
    }
    "adminpanel" -> "graphql" [
      likec4_id = "cloud.frontend.adminPanel:cloud.backend.graphql";
      style = "dashed";
      weight = 4;
      label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">fetches data in zero trust network<BR/>with sso authentification</FONT></TD></TR></TABLE>>;
    ];
    "dashboard" -> "graphql" [
      likec4_id = "cloud.frontend.dashboard:cloud.backend.graphql";
      style = "dashed";
      weight = 4;
      label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">fetches data</FONT></TD></TR></TABLE>>;
    ];
  }
  "support" -> "adminpanel" [
    likec4_id = "support:cloud.frontend.adminPanel";
    style = "dashed";
    weight = 2;
    label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">manages</FONT></TD></TR></TABLE>>;
  ];
  "customer" -> "dashboard" [
    likec4_id = "customer:cloud.frontend.dashboard";
    style = "dashed";
    weight = 2;
    label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">opens in browser</FONT></TD></TR></TABLE>>;
  ];
  "storage" -> "amazon" [
    likec4_id = "cloud.backend.storage:amazon";
    style = "dashed";
    weight = 2;
    label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">persists artifacts</FONT></TD></TR></TABLE>>;
    arrowhead = "normal";
    arrowtail = "odiamond";
    dir = "both";
    minlen = 0;
  ];
}"
`;

exports[`ElementViewPrinter > computedCloudView 1`] = `
"digraph {
  bgcolor = "transparent";
  layout = "dot";
  compound = true;
  rankdir = "TB";
  TBbalance = "min";
  splines = "spline";
  outputorder = "nodesfirst";
  nodesep = 1.528;
  ranksep = 1.667;
  pack = 120;
  packmode = "array_3";
  pad = 0.209;
  fontname = "Arial";
  graph [
    fontsize = 15;
    labeljust = "l";
    labelloc = "t";
    margin = 50.1;
  ];
  edge [
    arrowsize = 0.75;
    fontname = "Arial";
    fontsize = 14;
    penwidth = 2;
    color = "#6E6E6E";
    fontcolor = "#C6C6C6";
  ];
  node [
    fontname = "Arial";
    shape = "rect";
    width = 4.445;
    height = 2.5;
    style = "filled";
    penwidth = 0;
  ];
  "customer" [
    likec4_id = "customer";
    likec4_level = 0;
    fillcolor = "#3b82f6";
    fontcolor = "#eff6ff";
    color = "#2563eb";
    margin = "0.362,0.362";
    label = <<FONT POINT-SIZE="19">customer</FONT>>;
  ];
  "support" [
    likec4_id = "support";
    likec4_level = 0;
    fillcolor = "#3b82f6";
    fontcolor = "#eff6ff";
    color = "#2563eb";
    margin = "0.362,0.362";
    label = <<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="6"><TR><TD><FONT POINT-SIZE="19">Support Engineer</FONT></TD></TR><TR><TD><FONT POINT-SIZE="14" COLOR="#bfdbfe">Support engineers are responsible for<BR/>supporting customers</FONT></TD></TR></TABLE>>;
  ];
  "amazon" [
    likec4_id = "amazon";
    likec4_level = 0;
    fillcolor = "#3b82f6";
    fontcolor = "#eff6ff";
    color = "#2563eb";
    margin = "0.362,0.362";
    label = <<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="6"><TR><TD><FONT POINT-SIZE="19">Amazon</FONT></TD></TR><TR><TD><FONT POINT-SIZE="14" COLOR="#bfdbfe">Amazon is a cloud provider</FONT></TD></TR></TABLE>>;
  ];
  subgraph "cluster_cloud" {
    likec4_id = "cloud";
    likec4_level = 0;
    likec4_depth = 1;
    fillcolor = "#194b9e";
    color = "#1b3d88";
    style = "filled";
    margin = 40;
    label = <<FONT POINT-SIZE="11" COLOR="#bfdbfeb3"><B>CLOUD</B></FONT>>;
    "frontend" [
      likec4_id = "cloud.frontend";
      likec4_level = 1;
      fillcolor = "#3b82f6";
      fontcolor = "#eff6ff";
      color = "#2563eb";
      margin = "0.417,0.445";
      label = <<FONT POINT-SIZE="19">Frontend</FONT>>;
    ];
    "backend" [
      likec4_id = "cloud.backend";
      likec4_level = 1;
      fillcolor = "#3b82f6";
      fontcolor = "#eff6ff";
      color = "#2563eb";
      margin = "0.362,0.362";
      label = <<FONT POINT-SIZE="19">Backend</FONT>>;
    ];
    "frontend" -> "backend" [
      likec4_id = "cloud.frontend:cloud.backend";
      style = "dashed";
      weight = 4;
      label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14"><B>[...]</B></FONT></TD></TR></TABLE>>;
      minlen = 0;
    ];
  }
  "customer" -> "frontend" [
    likec4_id = "customer:cloud.frontend";
    style = "dashed";
    weight = 3;
    label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">opens in browser</FONT></TD></TR></TABLE>>;
  ];
  "support" -> "frontend" [
    likec4_id = "support:cloud.frontend";
    style = "dashed";
    weight = 3;
    label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">manages</FONT></TD></TR></TABLE>>;
  ];
  "backend" -> "amazon" [
    likec4_id = "cloud.backend:amazon";
    style = "dashed";
    weight = 2;
    label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">persists artifacts</FONT></TD></TR></TABLE>>;
    arrowhead = "normal";
    arrowtail = "odiamond";
    dir = "both";
    minlen = 0;
  ];
}"
`;

exports[`ElementViewPrinter > computedIndexView 1`] = `
"digraph {
  bgcolor = "transparent";
  layout = "dot";
  compound = true;
  rankdir = "TB";
  TBbalance = "min";
  splines = "spline";
  outputorder = "nodesfirst";
  nodesep = 1.528;
  ranksep = 1.667;
  pack = 120;
  packmode = "array_3";
  pad = 0.209;
  fontname = "Arial";
  graph [
    fontsize = 15;
    labeljust = "l";
    labelloc = "t";
    margin = 50.1;
  ];
  edge [
    arrowsize = 0.75;
    fontname = "Arial";
    fontsize = 14;
    penwidth = 2;
    color = "#6E6E6E";
    fontcolor = "#C6C6C6";
  ];
  node [
    fontname = "Arial";
    shape = "rect";
    width = 4.445;
    height = 2.5;
    style = "filled";
    penwidth = 0;
  ];
  "customer" [
    likec4_id = "customer";
    likec4_level = 0;
    fillcolor = "#3b82f6";
    fontcolor = "#eff6ff";
    color = "#2563eb";
    margin = "0.362,0.362";
    label = <<FONT POINT-SIZE="19">customer</FONT>>;
  ];
  "support" [
    likec4_id = "support";
    likec4_level = 0;
    fillcolor = "#3b82f6";
    fontcolor = "#eff6ff";
    color = "#2563eb";
    margin = "0.362,0.362";
    label = <<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="6"><TR><TD><FONT POINT-SIZE="19">Support Engineer</FONT></TD></TR><TR><TD><FONT POINT-SIZE="14" COLOR="#bfdbfe">Support engineers are responsible for<BR/>supporting customers</FONT></TD></TR></TABLE>>;
  ];
  "cloud" [
    likec4_id = "cloud";
    likec4_level = 0;
    fillcolor = "#3b82f6";
    fontcolor = "#eff6ff";
    color = "#2563eb";
    margin = "0.362,0.362";
    label = <<FONT POINT-SIZE="19">cloud</FONT>>;
  ];
  "amazon" [
    likec4_id = "amazon";
    likec4_level = 0;
    fillcolor = "#3b82f6";
    fontcolor = "#eff6ff";
    color = "#2563eb";
    margin = "0.362,0.362";
    label = <<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="6"><TR><TD><FONT POINT-SIZE="19">Amazon</FONT></TD></TR><TR><TD><FONT POINT-SIZE="14" COLOR="#bfdbfe">Amazon is a cloud provider</FONT></TD></TR></TABLE>>;
  ];
  "customer" -> "cloud" [
    likec4_id = "customer:cloud";
    style = "dashed";
    weight = 3;
    label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">opens in browser</FONT></TD></TR></TABLE>>;
  ];
  "support" -> "cloud" [
    likec4_id = "support:cloud";
    style = "dashed";
    weight = 3;
    label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">manages</FONT></TD></TR></TABLE>>;
  ];
  "cloud" -> "amazon" [
    likec4_id = "cloud:amazon";
    style = "dashed";
    weight = 3;
    label = <<TABLE BORDER="0" CELLPADDING="3" CELLSPACING="0" BGCOLOR="#18191bA0"><TR><TD ALIGN="TEXT" BALIGN="LEFT"><FONT POINT-SIZE="14">persists artifacts</FONT></TD></TR></TABLE>>;
    arrowhead = "normal";
    arrowtail = "odiamond";
    dir = "both";
    minlen = 0;
  ];
}"
`;
